### PR TITLE
ocamlPackages.lwd: init at 0.3

### LINKED ISSUE #136208: `lwd` ocaml library, and accompanying packages
**Project description**
`lwd` is a reactive DOM model library for ocaml, similar to what React is at its core. There are several packages that build upon `lwd`, and are part of the source repo itself (they seem to be analogous to existing libraries, but integrate `lwd`):

- `nottui`: renders user interface in the terminal
- `nottui-lwt`: add support for concurrent/asynchronous UI to Nottui 
- `nottui-pretty` is an interactive pretty-printer (based on [Pprint](https://github.com/fpottier/pprint))
- `tyxml-lwd`: is a Js_of_ocaml library for making interactive applications, using Lwd for updating the DOM and [Tyxml](https://github.com/ocsigen/tyxml) for writing typesafe HTML Document.

**Metadata**

* homepage URL: https://github.com/let-def/lwd
* source URL: https://github.com/let-def/lwd
* license: mit
* platforms: gnu/linux, darwin (as far as I know)

### PR DESCRIPTION
**Pull request by Alizter**
Added libraries
 - lwd
 - nottui
 - nottui-lwt
 - nottui-pretty
 - tyxml-lwd

Which are all part of the lwd sources. https://github.com/let-def/lwd

This fixes #136208.

cc @vbgl

Signed-off-by: Ali Caglayan <alizter@gmail.com>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by GabrielDougherty** [comment]
Result of `nixpkgs-review pr 221929` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>ocamlPackages.lwd</li>
    <li>ocamlPackages.nottui</li>
    <li>ocamlPackages.nottui-lwt</li>
    <li>ocamlPackages.nottui-pretty</li>
    <li>ocamlPackages.tyxml-lwd</li>
  </ul>
</details>

File: pkgs/development/ocaml-modules/lwd/nottui-lwt.nix [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+  inherit (lwd)
+    version
+    src
+    ;
+
+  minimalOCamlVersion = "4.08";
+  duneVersion = "3";
+
+  buildInputs = [ lwt lwd nottui notty seq ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/lwd/nottui-lwt.nix:22]:
```suggestion
  propagatedBuildInputs = [ lwt lwd nottui notty ];
```

File: pkgs/development/ocaml-modules/lwd/nottui-pretty.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  inherit (lwd)
+    version
+    src
+    ;
+
+  minimalOCamlVersion = "4.08";
+  duneVersion = "3";
+
+  buildInputs = [ lwd nottui notty seq ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/lwd/nottui-pretty.nix:21]:
```suggestion
  propagatedBuildInputs = [ nottui ];
```

File: pkgs/development/ocaml-modules/lwd/tyxml-lwd.nix [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  inherit (lwd)
+    version
+    src
+    ;
+
+  minimalOCamlVersion = "4.08";
+  duneVersion = "3";
+
+  buildInputs = [ js_of_ocaml js_of_ocaml-ppx lwd tyxml ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/lwd/tyxml-lwd.nix:21]:
```suggestion
  buildInputs = [ js_of_ocaml-ppx ];
  propagatedBuildInputs = [ js_of_ocaml lwd tyxml ];
```

File: pkgs/development/ocaml-modules/lwd/default.nix [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+  minimalOCamlVersion = "4.08";
+  duneVersion = "3";
+
+  src = fetchurl {
+    url = "https://github.com/let-def/lwd/releases/download/v${version}/lwd-${version}.tbz";
+    sha256 = "sha256-H/vyW2tn2OBuWwcmPs8NcINXgFe93MSxRd8dzeoXARI=";
+  };
+
+  buildInputs = [ seq ];
```
[the remark this answers is not in the document]
Comment (Alizter) [reply pkgs/development/ocaml-modules/lwd/default.nix:19]: When I do this, `seq` is no longer found by Dune.

[the remark this answers is not in the document]
Comment (vbgl) [reply pkgs/development/ocaml-modules/lwd/default.nix:19]: What do you mean? When moving `seq` from `buildInputs` to `propagatedBuildInputs` in the `lwd` package, then `lwd` fails to build and the error is what you wrote?

[the remark this answers is not in the document]
Comment (Alizter) [reply pkgs/development/ocaml-modules/lwd/default.nix:19]: When I do the change you suggest here, I get this:
```
nix-repl> :b ocamlPackages.lwd
error: builder for '/nix/store/vc1aj7f7i5m0q5zkxhcp3apb101h8lqz-ocaml4.14.1-lwd-0.3.drv' failed with exit code 1;
       last 10 log lines:
       > File "lib/lwd/dune", line 5, characters 13-16:
       > 5 |   (libraries seq)
       >                  ^^^
       > Error: Library "seq" not found.
       > -> required by library "lwd" in _build/default/lib/lwd
       > -> required by _build/default/META.lwd
       > -> required by _build/install/default/lib/lwd/META
       > -> required by _build/default/lwd.install
       > -> required by alias install
       >
       For full logs, run 'nix log /nix/store/vc1aj7f7i5m0q5zkxhcp3apb101h8lqz-ocaml4.14.1-lwd-0.3.drv'.

```

[the remark this answers is not in the document]
Comment (Alizter) [reply pkgs/development/ocaml-modules/lwd/default.nix:19]: Could it be that buildDunePackage is not correctly interacting with propogatedBuildInputs?

[the remark this answers is not in the document]
Comment (vbgl) [reply pkgs/development/ocaml-modules/lwd/default.nix]: Please double check the spelling of that attribute.

[the remark this answers is not in the document]
Comment (Alizter) [reply pkgs/development/ocaml-modules/lwd/default.nix]: Oh yes, you are correct, I cannot spell. :sob:

**Comment by GabrielDougherty** [comment]
Result of `nixpkgs-review pr 221929` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>ocamlPackages.lwd</li>
    <li>ocamlPackages.nottui</li>
    <li>ocamlPackages.nottui-lwt</li>
    <li>ocamlPackages.nottui-pretty</li>
    <li>ocamlPackages.tyxml-lwd</li>
  </ul>
</details>

**Comment by Alizter** [comment]
Thanks!

**Pull request merged**